### ISSUE TITLE
Unexpected (at least for me) behavior with GoogleLeNetTransformer

### ISSUE THREAD

**Issue by EderSantana** [sklearn-theano/sklearn-theano#66]
When transforming batches images with GoogleLeNetTransformer if the images have different sizes, I get features with different lengths. 240x320 images return the expected 4096 dimensional vectors, but 226x400 images returns a 6144 dimensional vectors. I assumed that the transformer reshaped all the images to a fixed size like OverFeat. I don't remember if the documentation said that we should resize the images ourselves.

**Comment by EderSantana** [comment]
I forgot to mention, I'm using output_layers=('pool5/7x7_s1',) as parameters to the transformer

**Comment by eickenberg** [comment]
Thanks @EderSantana for bringing this up. Differently shaped input should indeed lead to differently shaped output, at least in the `*Transformer`s. Any deviation from this behavior is arbitrary and needs to be chosen carefully. 

For OverFeat, the `Classifier` does indeed crop to a middle patch of the image. We implemented it that way, because the binary published does the same.

Caffe on the other hand does a *resizing* of the input image to fit the network input size, so a different preprocessing operation.

I am not sure to which policy we should adhere - do it similarly to the original, or do uniformly the same thing. What I see crystallizing is a base class for Transformers/Classifiers that provides both options, but that both are switched off for Transformers. For Classfiers, since we should probably have one value output by default, we should choose one of the strategies.

Alternatively, one could also have a preprocessing class that can be prepended in a `sklearn.pipeline.Pipeline` (modulo some [solved?] reshaping issues, modulo getting in other packages' [skdata] way?)

@kastnerkyle any input on this?

The relevant logic for extracting middle patches is [here](https://github.com/sklearn-theano/sklearn-theano/blob/master/sklearn_theano/feature_extraction/overfeat.py#L376) btw, in case you would like to apply it by hand in the mean time.

**Comment by kastnerkyle** [comment]
The transformers are designed so that applying to an image or batch of images > the minimum size effectively becomes dense feature extraction, useful for tasks like localization seen here: http://sklearn-theano.github.io/auto_examples/plot_multiple_localization.html . Tbh the current OverfeatLocalizer is a tiny wrapper over the transformed output, with some handy relabeling of classes. So dense feature extraction is useful for certain tasks as well.

On the Overfeat front, there needs to some time taken to figure out exactly what the Overfeat binary does. As I understand it, they took 5 crops (4 corners plus center) then did softmax over all predictions, but the current center crop has an advantage in speed. I could see adding an option "cropstyle" which is by default None, but also has string options for "center", "resize", and "multwindow", where multiwindow is what Overfeat does. This could be in both the Transformer and Classifier classes, since there are good reasons for doing it both ways (dense, resized, or over crops).

The problem I see with a preprocessor is resizing would densify the input since pipelines don't have "minibatch" processing. So I think it is reasonable to have helper functions which are linked into the class, but we would probably want to be sure they were standalone and did not introduce dependency on e.g. skimage. Minibatch processing is also something that is on my plate in the near future.

I am just now getting to where I could tackle this (maybe this week) but once again, any PR is welcome. A huge thanks for reporting this, and for using the library! Stress testing through usage is a great way to make sure it is useful for others :)

**Comment by EderSantana** [comment]
Surely, I think that having dense feature extraction is something useful. I just wanted to warn those that, like me, didn't know enough about that specific transformer. I was looking for something like Alex's net like features.

About batch preprocessing. This is necessary everywhere right? At least, AFAIK we can batch crop with numpy array slicing, but I don't know how to batch resize and interpolate/decimate images. If you have a nice and parallel solution for that, with a sklearn wrapper that would be awesome. Right now I'm using imagemagick in the original images. But I think I that a parallel `map` around spicy's `misc.imresize` would do the job. No need to add skimage for that.

**Comment by kastnerkyle** [comment]
Minibatch processing is on my list for the very near future, and
structuring that right would basically allow for resizing, preprocessing,
etc. per minibatch. But any function we write for that (using imresize or
some such) would be easy to use in this way - it just depends where in the
loop this is done. Batch center cropping is good because that reduces
memory pressure immediately, but I think multiwindowing and resizing should
be done on minibatch basis as they could both potentially result in a
memory increase.

Either way, the documentation should be clearer about the expected output,
and also some more effort on making input to the classifiers/transformer
consistent between them (uint8 vs. float32, 0-255 scaled vs. 0-1, etc.). I
am personally leaning toward uint8 since that is the most compact format
which would allow larger datasets in memory (also imread returns this
typically), but there is real utility in float32 0-1 scaled since that is
the most common format for general ML. Maybe even supporting both through
type inspection + throwing a warning with unexpected input (say float32
input *not* scaled between 0-1) would be the best way forward.

I am mainly thinking of the use case where either an hdf5, memmap, or
iterator to a bunch of files is used for processing. Any kind of dense
operation will blow up the memory immediately, but being able to classify
or extract down to feature vector of 4096 a big hdf5 of images (or having a
nice function to process a directory/glob of files as an iterator) could be
very useful for many tasks. At some point you can't even store the *output*
for that many files without dumping to disk, but that edge case can be
solved when it becomes an issue.
